File: Python/1034-head_of_a_gang.py
from collections import defaultdict, deque


def head_of_gang(graph, thre, edge_weights, head_weights):
    def bfs(vertex):
        max_h_weight, head = 0, None
        n_member, total_weight = 0, 0
        queue = deque([vertex])
        visited[vertex] = True
        while queue:
            u = queue.popleft()
            n_member += 1
            if head_weights[u] > max_h_weight:
                head = u
                max_h_weight = head_weights[u]
            for v in graph[u]:
                total_weight += edge_weights[(u, v)]
                if visited[v] is False:
                    queue.append(v)
                    visited[v] = True
        if n_member > 2 and total_weight > thre:
            return head, n_member
        else:
            return None
    heads = []
    visited = defaultdict(bool)
    for people in graph:
        if visited[people] is False:
            head = bfs(people)
            if head:
                heads.append(head)
    return sorted(heads)


if __name__ == '__main__':
    n_call, thre = [int(i) for i in input().split()]
    graph = defaultdict(set)
    head_weights = defaultdict(int)
    edge_weights = defaultdict(int)
    for _ in range(n_call):
        u, v, w = input().split()
        w = int(w)
        graph[u].add(v)
        graph[v].add(u)
        head_weights[u] += w
        head_weights[v] += w
        edge_weights[(u, v)] = w
    heads = head_of_gang(graph, thre, edge_weights, head_weights)
    print(len(heads))
    for head, n_member in heads:
        print(head, n_member)

File: C++/1034-head_of_a_gang.cc
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <set>
#include <queue>
#include <algorithm>
#include <utility>


// types
using Graph = std::map<std::string, std::set<std::string>>;
using Head = std::pair<std::string, int>;  // (head, n_member)
struct Gang {
    std::string head;
    int n_member, total_weight;
};


Graph graph;
std::map<std::string, int> head_weights;
std::map<std::string, int> edge_weights;
std::map<std::string, bool> visited;


Gang bfs(std::string u) {
    int max_h_height = 0;
    int n_member = 0, total_weight = 0;
    std::string head;
    std::queue<std::string> Queue;
    Queue.push(u);
    visited[u] = true;
    while (!Queue.empty()) {
        std::string u = Queue.front();
        Queue.pop();
        ++n_member;
        if (head_weights[u] > max_h_height) {
            head = u;
            max_h_height = head_weights[u];
        }
        for (auto &v : graph[u]) {
            total_weight += edge_weights[u + v];
            if (!visited[v]) {
                Queue.push(v);
                visited[v] = true;
            }
        }
    }
    return {head, n_member, total_weight};
}


std::vector<Head> head_of_gang(int thre) {
    std::vector<Head> heads;
    for (auto &kv : graph) {
        if (!visited[kv.first]) {
            Gang gang = bfs(kv.first);
            if (gang.n_member > 2 && gang.total_weight > thre) {
                heads.push_back(std::make_pair(gang.head, gang.n_member));
            }
        }
    }
    std::sort(heads.begin(), heads.end());
    return heads;
}


int main() {

    int n_call, thre;
    std::cin >> n_call >> thre;

    std::string u, v;
    int w;
    while (n_call--) {
        std::cin >> u >> v >> w;
        graph[u].insert(v);
        graph[v].insert(u);
        head_weights[u] += w;
        head_weights[v] += w;
        edge_weights[u + v] = w;
    }

    std::vector<Head> heads = head_of_gang(thre);
    std::cout << heads.size() << '\n';
    for (auto &x : heads) {
        std::cout << x.first << ' ' << x.second << '\n';
    }
    
    return 0;
}
